### ISSUE TITLE
Feature Request for Block Diagram with Network

### ISSUE THREAD

**Issue by johnkw2275** [almende/vis#758]
I use the Network feature to create block diagrams and it works really well.  There is one catch.  I would really like to show sub blocks within a block.  

I will likely try to do this myself but I am not great at SW so will likely take me a while and be messy.  So figured it would be worth posting in case it is very simple for someone else to do.  

The easier solution:
The tool already supports assigning groups to the blocks.  The goal would be that if a GROUP is assigned to a set of blocks then a larger block outline would be created that encompasses all of the blocks marked with that group.  Then the block would be labeled with the name of the group.  

The more difficult solution:
The tool already supports a display mode where as you zoom in or out it will compress nodes.  The idea would be to support this one step further were those nodes are predefined levels of a block diagram.  

![nested block diagram](https://cloud.githubusercontent.com/assets/11774977/6969433/3759bafc-d934-11e4-97dd-ee0539450a18.png)

**Comment by AlexDM0** [comment]
Hi,

You could dynamically create SVG images and use those as images for the nodes. I think that would do exactly what your picture shows. 

Regards,

Alex

**Comment by AlexDM0** [comment]
Maybe this example can help you out:
http://jsbin.com/hahubu/edit?html,output

Additionally we have an example on using html in network on our website. 

Regards,

Alex

**Comment by johnkw2275** [comment]
First I really appreciate the fast and detailed answers.  Very helpful!

Although this is a neat feature, the reason for doing this was to avoid
rendering to an image.  I could create all the block diagrams in various
drawing programs (which is how I used it in the past) but it made it much
more difficult to edit and manipulate in real time.
I love the features within vis/network that allow me to drag and drop
components, and a list of other features that make it really simple to use
for block diagrams.

I actually enter the block diagrams in a spread sheet format and then use
the spreadsheet to generate the JSON.  In an earlier version I stored the
spreadsheets in Google docs and used the javascript to do the JSON
formatting which I loved, except for the fact that it doesn't work in a few
minor use cases.

So I love this feature and may actually use it for other things, but not
necessarily accomplishing what I was hoping for in this case.

**Comment by AlexDM0** [comment]
Hi John,

The idea is that the rendering of the images is just a preprocessing step before loading it into the network. You could then use all the interactivity vis provides with the nodes that look like what you're referring to.

I reread your issue and I didn't think you were referring to the clustering initially. If you are referring to the clustering, the next release of vis (4.0) will have full user control over the clustering mechanism. It has been completely redone. To get the drawstyle you want though, you're probably doomed to generate the images manually anyway, but you can do this for a cluster and get the clustering / declustering mechanisms if required.

I'd be interested to hear why the solution I suggested would not be good for you.

Regards,

Alex

**Comment by johnkw2275** [comment]
I actually just saw the cluster idea this morning. I hadn't seen that
previously but I think that has the potential to do exactly what I want.

I may have oversimplified the block diagram a little but the idea in the
text was a high level block diagram would just show the Data, processor,
and display.
Then zooming in a little would show all the parts in the processor.  Then a
level deeper might show all the parts in the processing IF block.
The idea is at each step there are additional connections.

If I understand the other example I don't think I have the same freedom to
show all the different connections.  But I am pretty excited to use that
other tool for documentation.  It is a really useful tool and I had never
seen it before.

[image: Inline image 1]

**Comment by AlexDM0** [comment]
Hi John,

We just released v4 today. I'm curious to hear what you come up with using the new clustering system.

Cheerd

**Comment by AlexDM0** [comment]
Is this still an issue btw?

**Comment by johnkw2275** [comment]
It is still something missing for me.  I have been testing with the
solution at gojs.com and that supports the functionality that I was
interested in although I like the visjs solution better for a number of
reasons.

**Comment by AlexDM0** [comment]
Hi,

You mentioned you wanted to use the clustering but I'll assume that this did not work out? The rendering events will allow you to draw additional information on the canvas which you could use. Would that solve your problem?

Regards,

Alex

**Comment by johnkw2275** [comment]
Sorry, I was traveling when you responded and kind of lost track when I got
back.  There may be ways to get it to work with rendering events on the
canvas, however I am by no means a programmer.  So that is a little beyond
my skill sets.  I can do javascript and follow an API but not really sure
where to start trying to use rendering events.

The clustering didn't work exactly the way I was hoping it would.

Basically the functionality in gojs found here:
http://gojs.net/latest/samples/basic.html is the basic functionality I was
looking for.  Since May I had converted over to the free trial of this.  If
it works well then I may go this route.  I will say my vast preference
would be to stick with vis because I have a lot of previous experience with
it, and have a lot of previously developed tools that have been really
valuable.

Eventually if I switch to gojs it also has the following feature found here
that is interesting to me.  http://gojs.net/latest/samples/grouping.html

I had thought about trying to develop the features myself and offer them
back to VIS but I am afraid unfortunately I don't think I have the
expertise at this time.    But these two features are very powerful
enablers for people doing block diagrams.  I know it would apply to most
engineering and scientific disciplines.

I really do appreciate the contact and following up!

Thanks

Kevin
